refactor/handle-all-localization-project

ahmedNaser7 opened the pull request:
Refactor : Use UiText for handling string resources and apply it in SnackBar messages in MoreScreen

Refactor : add onRetry to baseViewModel

This change allows for more flexible message handling, accommodating both string resources and dynamic strings.

Key changes:
- Introduced `UiText.kt` with `DynamicString` and `StringResource` sealed classes.
- Updated `MoreViewModel` to use `UiText.StringResource` for error and success messages.
- Modified `MoreScreenState` to store `errorMessage` and `successMessageSnackBar` as `UiText?`.
- Updated `SnackBar` composable to accept an optional `uiText: String?` parameter and use it if provided, otherwise falling back to the existing `@StringRes text: Int` parameter.
- Updated `MoreScreen` to pass the string representation of `UiText` to the `SnackBar` composable.

- The onRetry callback is added to the `tryToExecute` function in `BaseViewModel` to allow retrying an operation after an error occurs.

ahmedNaser7 committed:
Refactor: Use UiText for handling string resources and apply it in SnackBar messages in MoreScreen

This commit refactors the MoreScreen and its associated ViewModel and State to utilize the newly introduced `UiText` sealed class for handling SnackBar messages.

This change allows for more flexible message handling, accommodating both string resources and dynamic strings.

Key changes:
- Introduced `UiText.kt` with `DynamicString` and `StringResource` sealed classes.
- Updated `MoreViewModel` to use `UiText.StringResource` for error and success messages.
- Modified `MoreScreenState` to store `errorMessage` and `successMessageSnackBar` as `UiText?`.
- Updated `SnackBar` composable to accept an optional `uiText: String?` parameter and use it if provided, otherwise falling back to the existing `@StringRes text: Int` parameter.
- Updated `MoreScreen` to pass the string representation of `UiText` to the `SnackBar` composable.

ahmedNaser7 committed:
fix : add onRetry to baseViewModel

The onRetry callback is added to the tryToExecute function in BaseViewModel to allow retrying an operation after an error occurs.

ahmedNaser7 committed:
Merge branch 'develop' into refactor/handle-all-localization-project

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/SnackBar.kt:
```diff
@@ -24,6 +24,7 @@ import com.paris_2.san3a.presentation.shared.utils.PreviewMultiDevices
 @Composable
 fun SnackBar(
     @StringRes text: Int,
+    uiText:String?=null,
```
why did you keep the text with the uiText ?

ahmedNaser7 replied:
i will delete it , and handle all screens that use  , but i need approve from the team

Joseph-Sameh-0 reviewed (commented):
I don't have a problem with this approach

mohammed-akkad commented on app/src/main/java/com/paris_2/san3a/presentation/utill/UiText.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Composable
+    fun asString(): String {
```
What is different between this functions?

Joseph-Sameh-0 replied:
one composable has context and one not composable needs context

ahmedNaser7 replied:
Function with context to handle the localization in 
Activity  in addittion to composable function

ahmedNaser7 committed:
feat: implement retry limit for coroutine execution

This commit introduces a retry limit functionality to the `tryToExecute` method in `BaseViewModel`.

The `tryToExecute` method now accepts an optional `retryLimit` parameter. If provided, the `onRetry` block will be executed up to the specified number of times in case of an exception during the coroutine execution.

ahmedNaser7 committed:
feat: implement retry limit for coroutine execution

This commit introduces a retry limit functionality to the `tryToExecute` method in `BaseViewModel`.

The `tryToExecute` method now accepts an optional `retryLimit` parameter. If provided, the `onRetry` block will be executed up to the specified number of times in case of an exception during the coroutine execution.

ahmedNaser7 committed:
Refactor: reorder parameters in BaseViewModel's safeApiCall function

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/handle-all-localization-project

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/more/moreScreen/MoreScreen.kt:
```diff
@@ -221,28 +221,30 @@ private fun MoreScreenContent(
                                 .align(Alignment.TopCenter),
-                            onClick = moreInteractionListener::onDismissSnackBar
+                            onClick = moreInteractionListener::onDismissSnackBar,
+                            text = 0
```
what is this 0?

ahmedNaser7 committed:
Refactor: remove UiText class

The UiText class, which was used for handling dynamic and string resource texts, has been removed. This functionality will be handled by other means.

ahmedNaser7 committed:
Refactor UiText for improved string resource handling

This commit refactors the `UiText` sealed class to provide more robust and flexible string resource management.

Key changes:
- Renamed `value` to `text` in `DynamicString` for clarity.
- Modified `StringResource` to accept `vararg args: Any` for formatted strings.
- Introduced `asString(context: Context)` function for resolving `UiText` instances outside of Composable contexts.
- Updated the Composable `asString()` function to leverage the new context-based `asString` and `handleStringResources`.
- Added a private `handleStringResources` function to encapsulate the logic for retrieving and formatting string resources, including basic error handling.

ahmedNaser7 committed:
Update SnackBar component to accept string directly

The SnackBar component has been updated to accept a `String` directly for its text content, instead of a `StringRes` ID. This simplifies its usage by removing the need for string resource resolution within the component.

ahmedNaser7 committed:
Refactor: Update VerificationScreen and ViewModel to use UiText for messages

This commit refactors the `VerificationScreen` and `VerificationViewModel` to utilize `UiText` for handling string resources, enabling better support for dynamic and localized text.

Specifically, the following changes were made:

-   In `VerificationViewModel`:
    -   `VerificationScreenState`'s `successMessageSnackBar` and `errorMessage` properties are now of type `UiText?` instead of `@StringRes Int?`.
    -   `VerificationUiState`'s `errorMessage` property is now of type `UiText?` instead of `@StringRes Int?`.
    -   Error and success messages are now set using `UiText.StringResource()`.
-   In `VerificationScreen`:
    -   The `text` parameter for `AppSnackBar` now uses `verificationScreenState.errorMessage.asString()` and `verificationScreenState.successMessageSnackBar.asString()` to display the messages.
    -   Placeholder `TODO()` calls for `onFrontOfNationalIdUploadClick` and `onBackOfNationalIdUploadClick` in the preview have been replaced with empty lambdas.

ahmedNaser7 committed:
Refactor: use UiText for error messages in MyJobsCraftsmanViewModel and related UI components

This commit updates `MyJobsCraftsmanViewModel` to use `UiText` for displaying error messages in SnackBars and other UI elements. This change provides better support for internationalization and improves the overall maintainability of the codebase.

The following files were also updated to reflect this change:
- `RequestDetails.kt`: Updated `OrderTrackItem` to use `UiText` for status messages.
- `MyOfferCraftsmanUiState.kt`: Updated `MyJobsCraftsmanScreenState` to use `UiText` for `errorMessage`.
- `MyJobsScreen.kt`: Updated `SnackBar` to display `UiText` error messages.

ahmedNaser7 committed:
Refactor: use UiText for error message in OTPRegisterScreen

The `errorMessage` field in `OTPRegisterScreenState` has been changed from `@StringRes Int?` to `UiText?`.

This change allows for more flexible error message handling, as `UiText` can represent both string resources and dynamic strings.

The `OTPRegisterScreen` and `OTPRegisterViewModel` have been updated to reflect this change.

ahmedNaser7 committed:
Refactor: Update error and success messages to use UiText

This commit refactors the `MyServiceViewModel` to use `UiText` for handling error and success messages.
This provides better support for internationalization and dynamic string resources.

The `MyServiceScreen` has been updated to correctly display the `UiText` messages using `asString()`.

ahmedNaser7 committed:
Refactor: Update UiText Usage and String Resources

This commit refactors the usage of `UiText` in `MoreViewModel`, `MoreScreenState`, `LocationViewModel`, and `LocationScreenState` to use the shared `UiText` class.

Additionally, it updates several composables in the `more.components` package (`LogoutItem`, `SettingItems`, `SettingItem`) to use `stringResource` for accessing string resources directly, removing the `@StringRes` annotation from parameters.

Finally, in `MoreScreen.kt`, the `SnackBar` composable's `uiText` parameter has been renamed to `text` and unnecessary `text = 0` assignments have been removed.

ahmedNaser7 committed:
Fix: pass string resource to SnackBar

The SnackBar composable expects a String, not an Int resource ID. This commit updates the calls to `SnackBar` to use `stringResource()` to correctly retrieve the string values.

ahmedNaser7 committed:
Organize imports in AccountScreen and AccountViewModel
